360 Panorama Viewer (front-end only)

c-probably-archived20260718 opened the pull request:
This adds basic front-end support for images with [spherical XMP metadata](https://developers.google.com/streetview/spherical-metadata) and partially resolves https://github.com/tootsuite/mastodon/issues/5867, but because the back-end strips all image metadata when uploading a media attachment (and because I don't know Ruby so I have no idea how to change that), this does not actually work without replacing the image files in the database.

![demo](https://user-images.githubusercontent.com/5412095/34461693-ac824e4c-ee31-11e7-9eb0-09093d31fde9.gif)

Some code adapted from [exif-js](https://github.com/exif-js/exif-js) and from [marzipano](https://github.com/google/marzipano).

The user interface is very basic and supports mouse-dragging, scroll-zooming, and pinch-zooming in Safari.

[Potential] issues:
- No back-end support
- Images are downsized a lot when uploading. Because the images are blown up when projected onto a sphere, this is much more noticeable
- accessibility?

c-probably-archived20260718 committed:
Add panorama viewer

c-probably-archived20260718 committed:
Add touch events in panorama viewer

c-probably-archived20260718 committed:
Only show panorama if there's GPano XMP data

c-probably-archived20260718 committed:
Merge https://github.com/tootsuite/mastodon

c-probably-archived20260718 committed:
Fix selector in comma sequence not being on its own single line

sorin-davidoi commented on app/javascript/mastodon/features/ui/components/image_loader.js:
```diff
@@ -1,6 +1,9 @@
 import classNames from 'classnames';
+import Motion from '../util/optional_motion';
+import spring from 'react-motion/lib/spring';
+import PanoramaViewer, { getPanoramaData } from './panorama_viewer';
```
Haven't looked at the size, but could this be lazy-loaded? Only a small fraction of images will be panoramas, so it does not really make sense to always load it (see https://github.com/tootsuite/mastodon/blob/master/app/javascript/mastodon/features/ui/util/async-components.js).

cpsdqs replied:
The part that checks for panoramas can't be lazy loaded, but the WebGL renderer could. Not sure if it'd be worth it though, the webpack chunks here are at least 30kB and the uncompressed renderer is just 9kB

cpsdqs replied:
oops I forgot about gl-matrix, which is huge. I can't find anything where the webpack chunks are defined though?

sorin-davidoi replied:
They are defined [here](https://github.com/tootsuite/mastodon/blob/master/app/javascript/mastodon/features/ui/util/async-components.js) - for an example how to use see [the emoji picker](https://github.com/tootsuite/mastodon/blob/master/app/javascript/mastodon/features/compose/components/emoji_picker_dropdown.js#L4).

c-probably-archived20260718 committed:
Lazy-load panorama viewer

c-probably-archived20260718 committed:
Clean up panorama viewer

Gargron commented:
This is really cool. But yeah I guess we need to figure out how to use ImageMagick's `-strip` without stripping out that spherical metadata.

c-probably-archived20260718 committed:
Respect initial(Heading|Pitch)Degrees, support ExifTool-generated XMP

cpsdqs commented:
Maybe the spherical metadata could be extracted before stripping and inserted back after?

akihikodaki commented:
XMP can contain any kind of data so it should be audited, but ImageMagick is not capable to do that.

> Maybe the spherical metadata could be extracted before stripping and inserted back after?

That is the only possible solution, I think.

Gargron commented:
>That is the only possible solution, I think.

Do we even have the tools to do that, or would it require adding another system-level dependency. Also, paperclip makes this very complicated :frowning:

akihikodaki commented:
> Do we even have the tools to do that, or would it require adding another system-level dependency.

Unfortunately there is no well-maintained Ruby implementation, so it would require system-level dependency. Exiv2, a command-line metadata editor, is available on Alpine Linux, Arch Linux, CentOS, Debian, Fedora, Homebrew, and RHEL.

> Also, paperclip makes this very complicated 😦

Not that much. Implementing a new processor by extending `Paperclip::Thumbnail` would be a good solution. See http://www.rubydoc.info/gems/paperclip/Paperclip/Thumbnail for the documentation.

cpsdqs commented:
can someone maybe add a “help wanted” tag?

akihikodaki commented:
#6317 deals with the help-wanted concern.

c-probably-archived20260718 committed:
Make dropdown animations respect their placement

Also fix the corner radius on the privacy dropdown button when using top placement

c-probably-archived20260718 committed:
Merge https://github.com/tootsuite/mastodon into pano

c-probably-archived20260718 committed:
Merge with upstream and put pano viewer in ZoomableImage

c-probably-archived20260718 committed:
Add scroll velocity and click handling in pano viewer

cpsdqs commented:
I just updated this, and commit [e7932c5](https://github.com/tootsuite/mastodon/pull/8292) is not part of this PR; I don’t know why that ended up there

c-probably-archived20260718 committed:
Add black background for cropped panoramas

neuhaus commented:
Looks like 99% of the work is already done here.. could we get this re-opened and merged?

Sasha-Sorokin commented:
This probably needs to be rebased @neuhaus. It's not very useful to reopen unless it is maintained, the PR may be recreated after all.
